src/validation.cpp

```diff
@@ -4670,20 +4681,48 @@ bool LoadGenesisBlock(const CChainParams& chainparams)
     return ::ChainstateActive().LoadGenesisBlock(chainparams);
 }
 
-void LoadExternalBlockFile(
+void LoadExternalBlockFiles(
+    const std::string& action,
     const CChainParams& chainparams,
-    FILE* fileIn,
-    FlatFilePos* dbp,
-    std::multimap<uint256, FlatFilePos>* blocks_with_unknown_parent)
+    const std::vector<fs::path>& blk_paths,
+    bool write_to_disk)
 {
-    assert((dbp && blocks_with_unknown_parent) || (!dbp && !blocks_with_unknown_parent));
+    std::multimap<uint256, FlatFilePos> blocks_with_unknown_parent;
+    for (size_t n_file = 0; n_file < blk_paths.size(); ++n_file) {
+        LogPrintf("%s block file %s ...\n", action, blk_paths[n_file].filename().string());
+        FILE* file = fsbridge::fopen(blk_paths[n_file], "rb");
+        if (file == nullptr) {
+            LogPrintf("%s: Warning: Could not open blocks file %s\n", __func__, blk_paths[n_file]);
+            return;
+        }
```

Comment:
Isn't this kinda duplication of what `ReadBlockFromDisk` does? I'm wondering why `FlatFilePos pos(n_file, 0);` isn't created, passed to  `ReadBlockFromDisk` (like before, when signature took `FlatFilePos`), and then you can pass it to `LoadExternalBlockFile`

Reply:
This duplication exists even without this PR. Before this PR, this `fopen()` happens [here](https://github.com/bitcoin/bitcoin/pull/20331/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dL729). 

To summarize why this happens in two places: There are two ways blocks are "loaded" into memory when reindexing (and, now, with this PR, when using `-loadblock`). First, a blocks file is opened ([here](https://github.com/bitcoin/bitcoin/pull/20331/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R4693)) and read from disk (using large reads) sequentially. At this point, we don't know the individual blocks' starting offsets, so we can't create a `FlatFilePos` to pass to `ReadBlockFromDisk()` as you suggest. Besides, doing these large reads is more efficient, typically many blocks per read -- plus a partial block at the end (since, again, the block boundaries are unknown). Once in memory, the block boundaries are discovered, and each block is deserialized. But the blocks in these blocks files are mostly out of order, so most of the blocks can't be accepted (have `AcceptBlock()` called on it) because its parent has not yet been seen.

When such a parent is later found, the child block is then _re-read_ from disk using `ReadBlockFromDisk()`, deserialized a second time, and then passed to `AcceptBlock()` (allowed because its parent is now present). In this case, we're reading exactly a single block, because we know where it starts (in the file) and its length (which is what a `FlatFilePos` encodes).